src/ast/pass_manager.h

```diff
@@ -1,64 +1,371 @@
+  [[noreturn]] static void fail(int type_id);
+
+  // put emplaces the the value into the map; it must not exist already. In
+  // general this should not be called manually, it should only be called by
+  // individual passes during execution (the `run` function below).
+  template <typename T>
+  void put(T &&t)
+  {
+    int type_id = TypeId<T>::type_id();
+    state_.emplace(type_id, std::make_unique<T>(std::move(t)));
+  }
+
+  // put can also provide an external reference for the map; its lifecycle is
+  // managed outside the scope of the passes. While passes can produce results
+  // that are references, this should be generally discouraged (and this
+  // capability should eventually be removed). Like `put`, this should not be
+  // called directly and should only be called by `run`, below.
+  template <typename T>
+  void put(T &t)
+  {
+    int type_id = TypeId<T>::type_id();
+    extern_state_.emplace(type_id, t);
+  }
+
+  // Will register the given type id and the name, which is resolvable using
+  // lookup_type below. This is called exclusively by `type_id` above.
+  static int add_type(std::string &&type_name)
+  {
+    int type_id = next_type_id_++;
+    type_names_[type_id] = std::move(type_name);
+    return type_id;
+  }
+
+  // Returns the underlying type for this type_id. Note that the type must
+  // exist, or this will have undefined behavior. This is called exclusively
+  // by the pass manager, when statically checking pass types.
+  static const std::string &lookup_type(int type_id)
+  {
+    return type_names_[type_id];
+  }
+
+  // Provide a short-hand to determine:
+  // - Whether the context contains an `ASTContext`, and,
+  // - Whether this `ASTContext` has error diagnostics.
+  //
+  // If both of the above are true, then false is returned. This is a
+  // convenience method provided for the `PassManager` below, as passes will
+  // stop when an error diagnostic has been generated.
+  bool ok() const;
+
+  static std::atomic<int> next_type_id_;
+  static std::unordered_map<int, std::string> type_names_;
+
+  std::unordered_map<int, std::unique_ptr<State>> state_;
+  std::unordered_map<int, std::reference_wrapper<State>> extern_state_;
+
+  friend class PassManager;
+  friend class Pass;
 };
 
-using PassFPtr = std::function<void(PassContext &)>;
+// State is the template class used for states that may be stored in a
+// PassContext. In order to consume or produce information in a given pass,
+// simply override this class with the desired type.
+//
+// For example:
+//
+//   class Variables : State<"variables"> {
+//     // my fields
+//   };
+//
+// The `Variables` class can then be stored in the pass context.
+template <PassContext::TypeName name>
+class State : public PassContext::State {
+public:
+  static std::string type_name()
+  {
+    return name.str();
+  }
+};
 
-// Base pass
+// Pass is a arbitrary pass.
+//
+// It is permitted to produce a single output, but may consume any number of
+// inputs.
```

Comment:
```suggestion
// It is permitted to produce a single output, but may consume multiple inputs.
```

Reply:
I've tweaked this comment to be more precise regarding inputs and outputs. PTAL.